sdks/go/examples/native_wordcap/nativepubsubio/native.go

```diff
@@ -0,0 +1,236 @@
+func (r *PubSubRead) ProcessElement(bf beam.BundleFinalization, rt *sdf.LockRTracker, _ []byte, emit func(beam.EventTime, []byte)) (sdf.ProcessContinuation, error) {
+	// Register finalization callback
+	bf.RegisterCallback(5*time.Minute, func() error {
+		for _, m := range r.ProcessedMessages {
+			m.Ack()
+		}
+		r.ProcessedMessages = []*pubsub.Message{}
+		return nil
+	})
+
+	// Initialize PubSub client if one has not been created already
+	if r.Client == nil {
+		client, err := pubsub.NewClient(context.Background(), r.ProjectID)
+		if err != nil {
+			return sdf.StopProcessing(), err
+		}
+		r.Client = client
+	}
+
+	for {
+		ok := rt.TryClaim(r.Subscription)
+		if !ok {
+			return sdf.ResumeProcessingIn(5 * time.Second), nil
+		}
+		sub := r.Client.Subscription(r.Subscription)
+		ctx, cFn := context.WithCancel(context.Background())
+
+		// Because emitters are not thread safe and synchronous Receive() behavior
+		// is deprecated, we have to collect messages in a goroutine and pipe them
+		// out through a channel.
+		messChan := make(chan *pubsub.Message, 1)
+		go func(sendch chan<- *pubsub.Message) {
+			err := sub.Receive(ctx, func(ctx context.Context, m *pubsub.Message) {
+				messChan <- m
+			})
+			if (err != nil) && (err != context.Canceled) {
+				log.Errorf(ctx, "error reading from PubSub: %v, stopping processing", err)
+				cFn()
+				close(messChan)
+			}
+		}(messChan)
+
+		// Give the goroutines time to start polling.
+		time.Sleep(5 * time.Second)
+
+		for {
+			select {
+			case m, ok := <-messChan:
+				if !ok {
+					log.Debug(context.Background(), "stopping bundle processing")
+					return sdf.StopProcessing(), nil
+				}
+				r.ProcessedMessages = append(r.ProcessedMessages, m)
+				emit(beam.EventTime(m.PublishTime.UnixMilli()), m.Data)
+			default:
+				log.Debug(context.Background(), "cancelling receive context, scheduling resumption")
+				cFn()
+				return sdf.ResumeProcessingIn(10 * time.Second), nil
+			}
+		}
+	}
+}
+
+// NativeRead reads messages from a PubSub topic in a streaming context, outputting
+// received messages as a PCollection of byte slices. If the provided subscription
+// name exists for the given topic, the DoFn will read from that subscription; otherwise,
+// a new subscription with the given subscription name will be created and read from.
+//
+// This feature is experimental and subject to change, including its behavior and function signature.
+// Please use the cross-language implementation Read() instead.
+func NativeRead(s beam.Scope, project, topic, subscription string) beam.PCollection {
+	s = s.Scope("pubsubio.NativeRead")
+
+	psRead, err := NewPubSubRead(context.Background(), project, topic, subscription)
+	if err != nil {
+		panic(err)
+	}
+	return beam.ParDo(s, psRead, beam.Impulse(s))
+}
+
+// PubSubWrite is a structural DoFn representing writes to a given PubSub topic.
+type PubSubWrite struct {
+	ProjectID string
+	Topic     string
+	Client    *pubsub.Client
```

Comment:
Unexport the fields that don't need to be serialized please.

Reply:
Done